### PR TITLE
Adds feature where if review requirement not met, then request

### PR DESCRIPTION
**Pull request by gfog-floqast**
Fixes #

## Proposed changes:

* Adds feature where if a review requirement is not met, then request the review. This is beneficial so that a developer does not need to look at the review requirement(s) to determine who to request reviews from in the case that review requirements are not met.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Testing instructions:

* Create `.github/required-review.yaml` similar to the following:
```
- name: Terraform
  paths:
    - '**/*.tf'
  teams:
    - any-of:
      - all-of:
        - team1
        - team2
      - '@user1'

- name: CODEOWNERS
  paths:
    - 'CODEOWNERS'
  teams:
    - team1
    - '@user2'

- name: .github
  paths:
    - '.github/**'
  teams:
    - all-of:
      - team1
      - team2
```
* Make updates to each of the files, as needed
* Observe that on first pass the necessary teams/users are requested

Tested this by deploying fork to my local github env and using settings above. 
* Observed initial failure on all requirements and observed required reviews were requested:
Screenshot 2023-05-11 at 10 59 30 AM
* Received review from team1 and observed CODEOWNERS check passed:
Screenshot 2023-05-11 at 11 02 43 AM
* Received review from user1 and observed Terraform check passed:
Screenshot 2023-05-11 at 11 03 51 AM
* Received review from team2 and observed .github check passed:
Screenshot 2023-05-11 at 11 04 54 AM
* Additionally received review from user2, which had no affect on status.

### CODE REVIEW

File: projects/github-actions/required-review/CHANGELOG.md
Comment (jeherve) [inline projects/github-actions/required-review/CHANGELOG.md]: Good news, you do not need to fill this in. This changelog will be automatically updated when we next release a new version of this action, based off the different changelog entries in the `changelog/` directory (where you've added your own changelog entry already, thank you).

```suggestion
```

File: projects/github-actions/required-review/src/request-review.js
Comment (jeherve) [inline projects/github-actions/required-review/src/request-review.js]: I see you've ignored Prettier here and across the file in general. Could you remove the exception and run it, so the style remains consistent with the other files in the action?

File: projects/github-actions/required-review/src/request-review.js
Comment (jeherve) [inline projects/github-actions/required-review/src/request-review.js]: I like that you extracted that here, that's clean. 👍

I wonder if we should make further use of that extraction in the existing logic in `projects/github-actions/required-review/src/team-members.js`?

**Review by jeherve — CHANGES_REQUESTED** [review]
First of all, thank you for the contribution!

I like that addition, I can see how useful it can be. However, since it can generate additional noise for maintainers since it will generate notifications and / or emails when they get added as reviewers, I think this option should be opt-in. I would consequently recommend that this only be available when specifying it in the action configuration.

What do you think of that?

File: projects/github-actions/required-review/src/request-review.js
Comment (gfog-floqast) [reply projects/github-actions/required-review/src/request-review.js]: yup, I can pull this out. Is there a file or location that is preferred for this? ie a helper file, or I can call it `get-username.js` and reference it in both files.

**Comment by gfog-floqast** [comment]
Thanks @jeherve, I'll look into making this feature opt-in

File: projects/github-actions/required-review/src/request-review.js
Comment (jeherve) [reply projects/github-actions/required-review/src/request-review.js]: > or I can call it get-username.js and reference it in both files

This seems like a good approach to me 👍

**Review by jeherve — DISMISSED** [review]
I've pushed a commit to address some linting errors, hopefully that helps a bit.

I think we're still missing a check for the new `request-reviews` action parameter you added, before we try to request reviews.

**Comment by gfog-floqast** [comment]
@jeherve Thanks for addressing those linting errors. Regarding the `// prettier-ignore` [comment](https://github.com/Automattic/jetpack/pull/30653#discussion_r1192526060), is there anything you would like me to do with it? I copied that from the [team-members file](https://github.com/Automattic/jetpack/pull/30653/files#diff-a2d2e2075dd774cef44c3db803766dcf96aeb8486b1580b61deb9ac28c144674L31) so not sure if this needs an update.

**Comment by jeherve** [comment]
> Regarding the `// prettier-ignore` [comment](https://github.com/Automattic/jetpack/pull/30653#discussion_r1192526060), is there anything you would like me to do with it? I copied that from the [team-members file](https://github.com/Automattic/jetpack/pull/30653/files#diff-a2d2e2075dd774cef44c3db803766dcf96aeb8486b1580b61deb9ac28c144674L31) so not sure if this needs an update.

I'll let @anomiex weigh in on that, since he worked on this part of the code in #29322

**Comment by gfog-floqast** [comment]
> missing a check for the new request-reviews action parameter

Can I get more clarification on this, I'm not sure where I should be looking. thx

**Comment by jeherve** [comment]
> Can I get more clarification on this, I'm not sure where I should be looking. thx

Sure thing. Here, you add a review request when needed:
https://github.com/Automattic/jetpack/pull/30653/files#diff-d99dd1c9d319cf9914b2111b19b7bb88771bcb4e4cee56ace33a4a95812bda3eR36

Here, you added a new Action parameter, `request-reviews`, so someone setting up the GitHub action in their own repo can decide to request reviews or not:
https://github.com/Automattic/jetpack/pull/30653/files#diff-b6d548c4bd130b59b0d693c1bd892f31aec433b2abfdd7067dc4f3f9ca5b031bR26

But you do not check for that parameter later on, so that parameter is useless right now. I would recommend that you check if that parameter is set before to call `requestReview` here:
https://github.com/Automattic/jetpack/pull/30653/files#diff-d99dd1c9d319cf9914b2111b19b7bb88771bcb4e4cee56ace33a4a95812bda3eR36

**Comment by gfog-floqast** [comment]
garr - missed it in my commit :D  - It's there now

File: projects/github-actions/required-review/src/requirement.js
Comment (anomiex) [inline projects/github-actions/required-review/src/requirement.js]: This is not the right place to add the requesting. If the configuration is requiring a review from "teamA or teamB", and someone from teamB already reviewed, there's no need to request a review from teamA. Also putting it here may result in redundant API calls if teamA is listed in multiple requirement blocks, and will almost certainly result in more API calls than are needed.

IMO the best thing to do would be, somewhere in
https://github.com/Automattic/jetpack/blob/8602a933da022394247f3dff851b8bec639ed6c9/projects/github-actions/required-review/src/main.js#L84-L88
add the list of needed teams for `r` to a `Set`. You'll need to figure out how to get just the needed teams, of course. For example, if review from both teamX and ( teamY or userBob ) is required, and there's already a review from teamY, you'd ideally only want to request teamX, not teamY again and not userBob either.

Then, just after
https://github.com/Automattic/jetpack/blob/8602a933da022394247f3dff851b8bec639ed6c9/projects/github-actions/required-review/src/main.js#L93-L96
if the new input is true and the set isn't empty, print a message along the lines of `Requesting reviews from ${ members of the set }` then pass the whole set to requestReviewer. Since the [request reviewers API endpoint](https://docs.github.com/en/rest/pulls/review-requests?apiVersion=2022-11-28#request-reviewers-for-a-pull-request) accepts multiple reviewers in one call, only one call should need to be made.

File: projects/github-actions/required-review/src/get-username.js
Comment (anomiex) [inline projects/github-actions/required-review/src/get-username.js]: Add caching in here please. If called twice for the same `user`, it shouldn't hit the API again.

File: projects/github-actions/required-review/src/request-review.js
Comment (anomiex) [inline projects/github-actions/required-review/src/request-review.js]: Is that a concern here? We do it in the other place because we're comparing the name from the config with the name returned on the review. But here I'd hope submitting "someusername" instead of "someUserName" would just do the right thing.

**Review by anomiex — DISMISSED** [review]
> > Regarding the `// prettier-ignore` [comment](https://github.com/Automattic/jetpack/pull/30653#discussion_r1192526060), is there anything you would like me to do with it? I copied that from the [team-members file](https://github.com/Automattic/jetpack/pull/30653/files#diff-a2d2e2075dd774cef44c3db803766dcf96aeb8486b1580b61deb9ac28c144674L31) so not sure if this needs an update.
> 
> I'll let @anomiex weigh in on that, since he worked on this part of the code in #29322

I don't like when prettier decides to wrap a backquoted string over multiple lines, in large part because vim doesn't handle it very well. So I tend to `prettier-ignore` when that happens.

File: projects/github-actions/required-review/src/requirement.js
Comment (gfog-floqast) [reply projects/github-actions/required-review/src/requirement.js]: Thanks for the review @anomiex  After going through the code I went this route due to how reviewers seemed to be grabbed first and then it looked like it worked back to the teams. This route seemed the most direct w/o making major changes to the rest of the code. With that said, I agree 💯 with limiting unneeded API calls and will look into reducing these as you're suggesting.

I'm wondering if it might be reasonable to say that review requests only occur on the first run? On that initial run I think it's safe to say that there would be no reviews therefor all required reviewers should be requested. Then as reviews are received, there shouldn't be the further need to request them again, as you point out. There could be a case where a review was removed (new commit push, etc), but thinking of how codeowners files works, it only requests reviews the first time iirc, which is the functionality I'm looking to introduce. Not sure if this would make the above any more straight forward.

Comment (anomiex) [reply projects/github-actions/required-review/src/requirement.js]: If someone pushes a commit to a PR that touches additional files, that might also trigger a need for more reviewers. 

For example, if someone from teamX says "oh, you also need to change this other thing over here" in their review, and that other thing happens to require review from teamZ.

Comment (gfog-floqast) [reply projects/github-actions/required-review/src/requirement.js]: yup - great point. 👍🏻

File: projects/github-actions/required-review/src/request-review.js
Comment (gfog-floqast) [reply projects/github-actions/required-review/src/request-review.js]: Tested, and yes `someUserName` works just as well as `someusername`. So guess there's really no reason to factor out [this code](https://github.com/Automattic/jetpack/pull/30653/files#diff-a2d2e2075dd774cef44c3db803766dcf96aeb8486b1580b61deb9ac28c144674L26) as here we can just use `const login = team.slice( 1 );` and not touch `team-members.js`.

If you agree, I'll just put that code back so I'm touching fewer files.

Comment (anomiex) [reply projects/github-actions/required-review/src/request-review.js]: Sounds good to me!

**Comment by gfog-floqast** [comment]
Please let me know what you think of this approach. The main issue that I'm running into is accounting for 1) a nested `all-of` inside an `any-of` along with 2) a nested `any-of` inside an `all-of`. I tried creating a way to only add the `alls` in 1) if the `any` was not satisfied, but this then caused issues in 2) where the `all` would not get added if the `any` _was_ satisfied. Without some sort of messy counter I'm not sure how best to handle this situation. 

Currently I'm thinking the most optimal solution would be to add the `all-of` to the list for review request and if it gets an unneeded request that would be better than having a needed request not getting added. I can also add a check to see if a review is already requested and then not request again. This would lower notifications, but may cause issues when a new file needed to be touched that included a reviewer from scenario 1) who would now be required again. 

I'm open to other suggestions of course, I started running in circles so thought best to reach out. This is my first time working with javascript so I may not be familiar with some other tricks that exist in this language for handing a situation like this.

File: projects/github-actions/required-review/src/requirement.js
Comment (gfog-floqast) [inline projects/github-actions/required-review/src/requirement.js]: This is suggestion, if y'all agree with this solution I would look to update `{ 'any-of': config.teams }` to just be config.teams and rework the related code.

**Comment by gfog-floqast** [comment]
Please let me know what you think of this approach. The main issue that I'm running into is accounting for 1) a nested `all-of` inside an `any-of` along with 2) a nested `any-of` inside an `all-of`. I tried creating a way to only add the `alls` in 1) if the `any` was not satisfied, but this then caused issues in 2) where the `all` would not get added if the `any` _was_ satisfied. Without some sort of messy counter I'm not sure how best to handle this situation.

Currently I'm thinking the most optimal solution would be to add the `all-of` to the list for review request and if it gets an unneeded request that would be better than having a needed request not getting added. I can also add a check to see if a review is already requested and then not request again. This would lower notifications, but may cause issues when a new file needed to be touched that included a reviewer from scenario 1) who would now be required again.

I'm open to other suggestions of course, I started running in circles so thought best to reach out. This is my first time working with javascript so I may not be familiar with some other tricks that exist in this language for handing a situation like this.

**Comment by anomiex** [comment]
I don't like the globals inside `requirements.js`. I think that trying to keep them straight is probably where you're getting confused.

The functions returned from `buildReviewerFilter()` currently return an array of found reviewers, and when things get to the top if the result is non-empty then the requirement is satisfied. Let's change that so the functions now return two values (as an object with two keys): the array of found reviewers and an array or set of needed teams. 

* The [simple string case](https://github.com/Automattic/jetpack/blob/1a69c56be4e87ecd85e728240727842b15c2405c/projects/github-actions/required-review/src/requirement.js#L33-L37) will either return a non-empty array of found reviewers and an empty array/set of needed teams, or an empty array of found reviewers and a 1-element array/set containing the passed-in team.
* The ["any-of"](https://github.com/Automattic/jetpack/blob/1a69c56be4e87ecd85e728240727842b15c2405c/projects/github-actions/required-review/src/requirement.js#L83-L88) and ["all-of"](https://github.com/Automattic/jetpack/blob/1a69c56be4e87ecd85e728240727842b15c2405c/projects/github-actions/required-review/src/requirement.js#L94-L99) cases would both calculate their arrays of found reviewers as they do now. Then if the found array is empty they will merge the needed-teams returned by all the sub-filters to return, otherwise they'll return an empty array/set for needed-teams.

Then I'd replace `isSatisfied()` with a `needsReviewsFrom()` method that returns the needed-teams set instead of a boolean. [The code in main.js](https://github.com/Automattic/jetpack/blob/a9d73d48c21feb033e209ff7fb1e8a7dd44ed5f4/projects/github-actions/required-review/src/main.js#L82-L90) would then look something like this
```js
} else {
	const neededForRequirement = await r.needsReviewsFrom( reviewers );
	core.endGroup();
	if ( neededForRequirement.size === 0 ) {
		core.info( `Requirement "${ r.name }" is satisfied by the existing reviews.` );
	} else {
		core.error( `Requirement "${ r.name }" is not satisfied by the existing reviews.` );
		// Merge into `neededTeams` set for possible use after the loop.
		neededForRequirement.forEach( neededTeams.add, neededTeams );
	}
}
```
Instead of `ok`, before the loop we'd initialize `const neededTeams = new Set()` and the post-loop check would look for `neededTeams.size === 0`.

**Comment by gfog-floqast** [comment]
@jeherve @anomiex I've made updates to address the comments, please let me know if there are any further changes or additions you would like to see. Thanks.

File: projects/github-actions/required-review/changelog/request-review-if-not-satisfied
Comment (anomiex) [inline projects/github-actions/required-review/changelog/request-review-if-not-satisfied]:
```suggestion
Option to request review if requirement is not satisfied.
```

File: projects/github-actions/required-review/src/request-review.js
Comment (anomiex) [inline projects/github-actions/required-review/src/request-review.js]: I'd recommend doing it like this instead
```suggestion
	for ( const t of teams ) {
```

File: projects/github-actions/required-review/src/requirement.js
Comment (anomiex) [inline projects/github-actions/required-review/src/requirement.js]: The way you have the code above written, every element of `requirementsMet` will have `.length > 0`. So as long as `requirementsMet.length > 0`, we're good here.

File: projects/github-actions/required-review/src/requirement.js
Comment (anomiex) [inline projects/github-actions/required-review/src/requirement.js]: You changed `printSet` so this will work:
```suggestion
			return printSet(
				`${ indent }Members of ${ team }:`,
				teamReviewers,
				neededTeams
			);
```

File: projects/github-actions/required-review/src/requirement.js
Comment (anomiex) [inline projects/github-actions/required-review/src/requirement.js]: You know at this point that `neededTeams` is empty, so no need to run it through Set.
```suggestion
			return printSet( `${ indent }=>`, [ ...new Set( requirementsMet.flat( 1 ) ) ], [] );
```

**Review by anomiex — COMMENTED** [review]
Looks pretty good. I left a few comments inline, and I'll kick off the CI jobs where the linter will have some additional comments.

**Comment by gfog-floqast** [comment]
@anomiex Addressed comments and committed suggestions - I didn't get the lint results before committing suggested changes so I lost those runs if you could please retrigger those I'll make any needed updates.

**Comment by gfog-floqast** [comment]
@jeherve @anomiex could one of y'all re-trigger the workflows for me? please and thank you.

**Comment by gfog-floqast** [comment]
@jeherve Linting errors should be fixed and ready for a final 🤞🏻 run.

File: projects/github-actions/required-review/src/request-review.js
Comment (anomiex) [inline projects/github-actions/required-review/src/request-review.js]:
```suggestion
 * @param {string[]} teams - GitHub team slug, or @ followed by a GitHub user name.
```

**Comment by gfog-floqast** [comment]
Thanks @anomiex I committed your suggestion, also pushed up final linting error corrections.

**Review by anomiex — APPROVED** [review]
If CI is happy, let's go for it. Thanks!

**Pull request merged**
